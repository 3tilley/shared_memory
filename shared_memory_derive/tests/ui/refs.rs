//! References and trait objects are NOT allowed in shared memory

use shared_memory::SharedMemCast;

#[derive(SharedMemCast)]
pub struct MyStruct<'a> {
    x: &'a i32,
}

#[derive(SharedMemCast)]
pub struct MyStruct2<'a>(&'a i32);

#[derive(SharedMemCast)]
pub struct MyStructTraitObject<'a> {
    x: &'a dyn SharedMemCast,
}

#[derive(SharedMemCast)]
pub struct MyStruct2TraitObject<'a>(&'a dyn SharedMemCast);

// No lifetime on struct itself so forced to reject based on field
#[derive(SharedMemCast)]
pub struct MyStructStatic<T: SharedMemCast + 'static> {
    x: &'static T,
}

#[derive(SharedMemCast)]
pub struct MyStruct2Static<T: SharedMemCast + 'static>(&'static T);

fn main() {}
