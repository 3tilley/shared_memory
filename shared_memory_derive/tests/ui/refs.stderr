error[E0277]: the trait bound `&'a i32: shared_memory::cast::SharedMemCast` is not satisfied
 --> $DIR/refs.rs:7:8
  |
7 |     x: &'a i32,
  |        ^
  |        |
  |        the trait `shared_memory::cast::SharedMemCast` is not implemented for `&'a i32`
  |        help: consider removing 1 leading `&`-references
  |
  = help: the following implementations were found:
            <i32 as shared_memory::cast::SharedMemCast>
  = note: required by `shared_memory::cast::AssertIsSharedMemCast`

error[E0277]: the trait bound `&'a i32: shared_memory::cast::SharedMemCast` is not satisfied
  --> $DIR/refs.rs:11:26
   |
11 | pub struct MyStruct2<'a>(&'a i32);
   |                          ^
   |                          |
   |                          the trait `shared_memory::cast::SharedMemCast` is not implemented for `&'a i32`
   |                          help: consider removing 1 leading `&`-references
   |
   = help: the following implementations were found:
             <i32 as shared_memory::cast::SharedMemCast>
   = note: required by `shared_memory::cast::AssertIsSharedMemCast`

error[E0277]: the trait bound `&'a (dyn shared_memory::cast::SharedMemCast + 'a): shared_memory::cast::SharedMemCast` is not satisfied
  --> $DIR/refs.rs:15:8
   |
15 |     x: &'a dyn SharedMemCast,
   |        ^
   |        |
   |        the trait `shared_memory::cast::SharedMemCast` is not implemented for `&'a (dyn shared_memory::cast::SharedMemCast + 'a)`
   |        help: consider removing 1 leading `&`-references
   |
   = note: required by `shared_memory::cast::AssertIsSharedMemCast`

error[E0277]: the trait bound `&'a (dyn shared_memory::cast::SharedMemCast + 'a): shared_memory::cast::SharedMemCast` is not satisfied
  --> $DIR/refs.rs:19:37
   |
19 | pub struct MyStruct2TraitObject<'a>(&'a dyn SharedMemCast);
   |                                     ^
   |                                     |
   |                                     the trait `shared_memory::cast::SharedMemCast` is not implemented for `&'a (dyn shared_memory::cast::SharedMemCast + 'a)`
   |                                     help: consider removing 1 leading `&`-references
   |
   = note: required by `shared_memory::cast::AssertIsSharedMemCast`

error[E0277]: the trait bound `&'static T: shared_memory::cast::SharedMemCast` is not satisfied
  --> $DIR/refs.rs:24:8
   |
24 |     x: &'static T,
   |        ^
   |        |
   |        the trait `shared_memory::cast::SharedMemCast` is not implemented for `&'static T`
   |        help: consider removing 1 leading `&`-references
   |
   = note: required by `shared_memory::cast::AssertIsSharedMemCast`

error[E0277]: the trait bound `&'static T: shared_memory::cast::SharedMemCast` is not satisfied
  --> $DIR/refs.rs:28:56
   |
28 | pub struct MyStruct2Static<T: SharedMemCast + 'static>(&'static T);
   |                                                        ^
   |                                                        |
   |                                                        the trait `shared_memory::cast::SharedMemCast` is not implemented for `&'static T`
   |                                                        help: consider removing 1 leading `&`-references
   |
   = note: required by `shared_memory::cast::AssertIsSharedMemCast`

For more information about this error, try `rustc --explain E0277`.
